src/util/valuetransformer.cpp

```diff
@@ -1,35 +1,130 @@
 #include "util/valuetransformer.h"
 
-ValueTransformer::ValueTransformer() {
-}
+#include <QDomElement>
+#include <QDomNode>
+#include <QDomNodeList>
+#include <memory>
+#include <numeric>
+
+#include "skin/legacy/skincontext.h"
+
+class TransformNode {
+  public:
+    TransformNode() {
+    }
+    virtual ~TransformNode() {
+    }
+
+    virtual double transform(double argument) const = 0;
+    virtual double transformInverse(double argument) const = 0;
+};
+
+namespace {
+
+class TransformAdd : public TransformNode {
+  public:
+    TransformAdd(double addend)
+            : m_addend(addend) {
+    }
+
+    double transform(double argument) const override {
+        return argument + m_addend;
+    }
+
+    double transformInverse(double argument) const override {
+        return argument - m_addend;
+    }
+
+  private:
+    double m_addend;
+};
+
+class TransformInvert : public TransformNode {
+  public:
+    double transform(double argument) const override {
+        return -argument;
+    }
+
+    double transformInverse(double argument) const override {
+        return -argument;
+    }
+};
+
+class TransformNot : public TransformNode {
+  public:
+    double transform(double argument) const override {
+        return !static_cast<bool>(argument);
+    }
 
-void ValueTransformer::addTransformer(TransformNode* pTransformer) {
-    m_transformers.append(pTransformer);
+    double transformInverse(double argument) const override {
+        return !static_cast<bool>(argument);
+    }
+};
+
+class TransformIsEqual : public TransformNode {
+  public:
+    TransformIsEqual(double compareValue)
+            : m_compareValue(compareValue) {
+    }
+
+    double transform(double argument) const override {
+        return argument == m_compareValue;
+    }
+
+    double transformInverse(double argument) const override {
+        if (argument > 0.0) {
+            return m_compareValue;
+        }
+        return 0.0;
+    }
+
+  private:
+    double m_compareValue;
+};
+
+} // namespace
+
+void ValueTransformer::addTransformer(std::unique_ptr<TransformNode> pTransformer) {
+    m_transformers.push_back(std::move(pTransformer));
 }
 
 double ValueTransformer::transform(double argument) const {
-    foreach (const TransformNode* pNode, m_transformers) {
-        argument = pNode->transform(argument);
-    }
-    return argument;
+    return std::accumulate(m_transformers.cbegin(),
+            m_transformers.cend(),
+            argument,
+            [&](double argument, const auto& pNode) {
+                return pNode->transform(argument);
+            });
 }
 
 double ValueTransformer::transformInverse(double argument) const {
-    for (int i = m_transformers.size() - 1; i >= 0; --i) {
-        const TransformNode* pNode = m_transformers[i];
-        argument = pNode->transformInverse(argument);
-    }
-    return argument;
+    return std::accumulate(m_transformers.crbegin(),
```

Comment:
The semantic of ``std::accumulate`` does not fit here:
> Computes the sum of the given value ...
https://en.cppreference.com/w/cpp/algorithm/accumulate

In addition a range based loop is much easier to read. 
Please adjust transform above as well.

Reply:
Well, it does a reduce in a well defined order. `std::reduce` rearranges the order and would be unsuitable here. IMO it fits perfectly compared to the eyesore that is the reverse for-loop. I will see if I can make it work with a ranges `reverse view` and a ranged for loop, but I don't want to revert to the index-based for loop again.

Reply:
`std::ranges::reverse_view` suits. Hopefully old compilers doe not bite us. Else a `rbegin()` solution is also fine.

Reply:
right, so in case `reverse_view` doesn't work out, you'd prefer a for-loop over reverse iterators?

Reply:
No, I would prefer the reverse iterators in that case.

Reply:
Whoops, suboptimal wording. I was asking whether you prefer
```cpp
for (auto it = m_transformers.crbegin(); it != m_transformers.crend(); it++) {
    // etc
};
```
instead of the `std::accumulate` over reverse iterators.

Reply:
yeah, so its macOS again... please suggest your preferred alterative.

Reply:
The first with the reverse iterators. This is IMHO the most describtive version.

Reply:
I don't particularly agree because we're now just copy-pasting the [`std::accumulate` implementation](https://en.cppreference.com/w/cpp/algorithm/accumulate#Possible_implementation) into there, but whatever you prefer...